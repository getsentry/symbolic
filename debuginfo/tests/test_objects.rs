use std::fmt;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_debuginfo::{DebugSession, Function, Object, SymbolMap};

/// Helper to create neat snapshots for symbol tables.
struct SymbolsDebug<'a>(&'a SymbolMap<'a>);

impl fmt::Debug for SymbolsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for symbol in self.0.iter() {
            writeln!(
                f,
                "{:>16x} {}",
                &symbol.address,
                &symbol.name().unwrap_or("<unknown>")
            )?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for function trees.
struct FunctionsDebug<'a>(&'a [Function<'a>], usize);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for function in self.0 {
            writeln!(
                f,
                "\n{:indent$}> {:#x}: {} ({:#x})",
                "",
                function.address,
                function.name,
                function.size,
                indent = self.1 * 2
            )?;

            for line in &function.lines {
                writeln!(
                    f,
                    "{:indent$}  {:#x}: {}:{} ({})",
                    "",
                    line.address,
                    line.file.name_str(),
                    line.line,
                    line.file.dir_str(),
                    indent = self.1 * 2
                )?;
            }

            write!(f, "{:?}", FunctionsDebug(&function.inlinees, self.1 + 1))?;
        }

        Ok(())
    }
}

#[test]
fn test_breakpad() -> Result<(), Error> {
    // Using the windows version here since it contains all record kinds
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Breakpad(
    BreakpadObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        name: "crash.pdb",
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_breakpad_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("breakpad_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_breakpad_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!(
        "breakpad_functions",
        FunctionsDebug(&functions[..10], 0)
    );

    Ok(())
}

#[test]
fn test_elf_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_debug() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_symbols() -> Result<(), Error> {
    // TODO(ja): Why does crash.debug not retain the symbol table but report has_symbols
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("elf_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_elf_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("elf_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_mach_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_dsym() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("mach_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_mach_functions() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("mach_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pe_32() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
        ),
        arch: X86,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_pe_64() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/CrashWithException.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5c9e09599000)
        ),
        debug_id: DebugId {
            uuid: "f535c5fb-2ae8-4bb8-aa20-6c30be566c5a",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\Users\\sentry\\source\\repos\\CrashWithException\\x64\\Release\\CrashWithException.pdb"
        ),
        arch: Amd64,
        kind: Executable,
        load_address: 0x140000000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

// NB: No test for PE symbols because our executable does not export any symbols
// NB: No test for PE functions because we can only read debug info from PDBs

#[test]
fn test_pdb() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pdb(
    PdbObject {
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        load_address: 0x0,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_pdb_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("pdb_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_pdb_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("pdb_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}
