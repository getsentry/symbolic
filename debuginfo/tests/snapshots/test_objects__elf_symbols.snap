---
created: "2019-02-27T13:05:25.490529Z"
creator: insta@0.6.3
source: debuginfo/tests/test_objects.rs
expression: SymbolsDebug(&symbols)
---
            1558 _init
            1900 _ZN15google_breakpad13PageAllocator7FreeAllEv.isra.6
            194a _ZN15google_breakpad17ProcCpuInfoReader14GetValueAndLenEPm.isra.20.part.21
            196a _ZN15google_breakpad10TypedMDRVAI14MDRawDirectoryE9CopyIndexEjPS1_.isra.32.part.33
            198a _ZN15google_breakpad10TypedMDRVAI14MDRawDirectoryE9CopyIndexEjPS1_.isra.32
            19a8 _ZN15google_breakpad10TypedMDRVAIjE20CopyIndexAfterObjectEjPKvm.isra.34.part.35
            19c8 _ZN12_GLOBAL__N_114MinidumpWriterC2EPKciPKN15google_breakpad16ExceptionHandler12CrashContextERKNSt7__cxx114listINS3_12MappingEntryESaISA_EEERKNS9_INS3_9AppMemoryESaISF_EEEbmbPNS3_11LinuxDumperE.part.93
            19e8 _ZN12_GLOBAL__N_114MinidumpWriterD2Ev.constprop.123
            1a14 _ZN12_GLOBAL__N_114MinidumpWriter9WriteFileEP20MDLocationDescriptorPKc.constprop.120
            1c00 _ZN12_GLOBAL__N_114MinidumpWriter13WriteProcFileEP20MDLocationDescriptoriPKc.constprop.119
            1c70 main
            1dc0 _start
            1df0 deregister_tm_clones
            1e30 register_tm_clones
            1e70 __do_global_dtors_aux
            1e90 frame_dummy
            1ec0 _ZN12_GLOBAL__N_18callbackERKN15google_breakpad18MinidumpDescriptorEPvb
            1f00 _ZN15google_breakpad18MinidumpDescriptorD1Ev
            1f40 _ZN15google_breakpad16ExceptionHandler21InstallHandlersLockedEv
            2070 _ZN15google_breakpad16ExceptionHandler21RestoreHandlersLockedEv
            20f0 _ZN15google_breakpad16ExceptionHandlerD1Ev
            2440 _ZN15google_breakpad16ExceptionHandler25SendContinueSignalToChildEv
            2520 _ZN15google_breakpad16ExceptionHandler12GenerateDumpEPNS0_12CrashContextE
            29f0 _ZN15google_breakpad16ExceptionHandler12HandleSignalEiP9siginfo_tPv
            2bd0 _ZN15google_breakpad16ExceptionHandler13SignalHandlerEiP9siginfo_tPv
            2dc0 _ZN15google_breakpad16ExceptionHandler22SimulateSignalDeliveryEi
            2e40 _ZN15google_breakpad16ExceptionHandler21WaitForContinueSignalEv
            2f30 _ZN15google_breakpad16ExceptionHandler6DoDumpEiPKvm
            2fe0 _ZN15google_breakpad16ExceptionHandler11ThreadEntryEPv
            3070 _ZN15google_breakpad16ExceptionHandler13WriteMinidumpEv
            3210 _ZN15google_breakpad16ExceptionHandler14AddMappingInfoERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPKhmmm
            3300 _ZN15google_breakpad16ExceptionHandler17RegisterAppMemoryEPvm
            33a0 _ZN15google_breakpad16ExceptionHandler19UnregisterAppMemoryEPv
            3400 _ZN15google_breakpad16ExceptionHandler21WriteMinidumpForChildEiiRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPFbRKNS_18MinidumpDescriptorEPvbESC_
            3660 _ZN15google_breakpad30SetFirstChanceExceptionHandlerEPFbiPvS0_E
            3670 _ZN15google_breakpad16ExceptionHandlerC1ERKNS_18MinidumpDescriptorEPFbPvEPFbS3_S4_bES4_bi
            3a10 _ZN15google_breakpad16ExceptionHandler13WriteMinidumpERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPFbRKNS_18MinidumpDescriptorEPvbESC_
            3cb0 _ZNSt6vectorIPN15google_breakpad16ExceptionHandlerESaIS2_EE19_M_emplace_back_auxIIS2_EEEvDpOT_
            3da0 _ZN15google_breakpad18MinidumpDescriptorC1ERKS0_
            3f20 _ZN15google_breakpad18MinidumpDescriptor10UpdatePathEv
            42f0 _ZN15google_breakpad18MinidumpDescriptoraSERKS0_
            43c0 _ZN6logger5writeEPKcm
            4400 _ZN15google_breakpad14WriteMicrodumpEiPKvmRKNSt7__cxx114listINS_12MappingEntryESaIS4_EEEbmbRKNS_18MicrodumpExtraInfoE
            6e10 _ZNSt6vectorIhN15google_breakpad16PageStdAllocatorIhEEE15_M_range_insertIPKhEEvN9__gnu_cxx17__normal_iteratorIPhS3_EET_SB_St20forward_iterator_tag
            71f0 _ZN15google_breakpad11LinuxDumper8LateInitEv
            7200 _ZN15google_breakpad11LinuxDumper17EnumerateMappingsEv
            7ac0 _ZN15google_breakpad11LinuxDumperC1EiPKc
            8300 _ZN15google_breakpad11LinuxDumperD2Ev
            8370 _ZN15google_breakpad11LinuxDumperD0Ev
            8390 _ZNK15google_breakpad11LinuxDumper20GetCrashSignalStringEv
            8590 _ZNK15google_breakpad11LinuxDumper22GetMappingAbsolutePathERKNS_11MappingInfoEPc
            85e0 _ZN15google_breakpad12_GLOBAL__N_113ElfFileSoNameERKNS_11LinuxDumperERKNS_11MappingInfoEPcm.constprop.55
            8770 _ZNK15google_breakpad11LinuxDumper26HandleDeletedFileInMappingEPc.part.12.constprop.56
            8920 _ZN15google_breakpad11LinuxDumper30GetMappingEffectiveNameAndPathERKNS_11MappingInfoEPcmS4_m
            8a10 _ZN15google_breakpad11LinuxDumper8ReadAuxvEv
            8b30 _ZN15google_breakpad11LinuxDumper4InitEv
            8b70 _ZN15google_breakpad11LinuxDumper24StackHasPointerToMappingEPKhmmRKNS_11MappingInfoE
            8be0 _ZNK15google_breakpad11LinuxDumper11FindMappingEPKv
            8c30 _ZN15google_breakpad11LinuxDumper12GetStackInfoEPPKvPmm
            8cb0 _ZNK15google_breakpad11LinuxDumper17FindMappingNoBiasEm
            8d00 _ZN15google_breakpad11LinuxDumper17SanitizeStackCopyEPhmmm
            8fa0 _ZNK15google_breakpad11LinuxDumper26HandleDeletedFileInMappingEPc
            9000 _ZN15google_breakpad11LinuxDumper27ElfFileIdentifierForMappingERKNS_11MappingInfoEbjRNS_15wasteful_vectorIhEE
            9350 _ZN15google_breakpad13PageAllocator5AllocEm
            94a0 _ZNK15google_breakpad17LinuxPtraceDumper12IsPostMortemEv
            94b0 _ZN15google_breakpad17LinuxPtraceDumper15CopyFromProcessEPviPKvm
            9640 _ZN15google_breakpad17LinuxPtraceDumper13ThreadsResumeEv
            9700 _ZNK15google_breakpad17LinuxPtraceDumper13BuildProcPathEPciPKc
            97c0 _ZN15google_breakpad17LinuxPtraceDumper16EnumerateThreadsEv
            9e50 _ZN15google_breakpad17LinuxPtraceDumperC1Ei
            9e80 _ZN15google_breakpad17LinuxPtraceDumper15ReadRegisterSetEPNS_10ThreadInfoEi
            9f70 _ZN15google_breakpad17LinuxPtraceDumper13ReadRegistersEPNS_10ThreadInfoEi
            a050 _ZN15google_breakpad17LinuxPtraceDumper20GetThreadInfoByIndexEmPNS_10ThreadInfoE
            a690 _ZN15google_breakpad17LinuxPtraceDumper14ThreadsSuspendEv
            a8f0 _ZN15google_breakpad17LinuxPtraceDumperD1Ev
            a910 _ZN15google_breakpad17LinuxPtraceDumperD0Ev
            a930 _ZNSt6vectorIiN15google_breakpad16PageStdAllocatorIiEEE17_M_default_appendEm
            abc0 _ZN12_GLOBAL__N_114MinidumpWriter21WriteThreadListStreamEP14MDRawDirectory.constprop.105
            bb80 _ZN12_GLOBAL__N_114MinidumpWriter4DumpEv.constprop.104
            de00 _ZN15google_breakpad13WriteMinidumpEPKcRKNSt7__cxx114listINS_12MappingEntryESaIS4_EEERKNS3_INS_9AppMemoryESaIS9_EEEPNS_11LinuxDumperE
            e370 _ZN15google_breakpad13WriteMinidumpEPKcii
            ea30 _ZN12_GLOBAL__N_117WriteMinidumpImplEPKciliPKvmRKNSt7__cxx114listIN15google_breakpad12MappingEntryESaIS7_EEERKNS5_INS6_9AppMemoryESaISC_EEEbmb
            f0c0 _ZN15google_breakpad13WriteMinidumpEPKciPKvmbmb
            f1e0 _ZN15google_breakpad13WriteMinidumpEiiPKvmbmb
            f300 _ZN15google_breakpad13WriteMinidumpEPKciPKvmRKNSt7__cxx114listINS_12MappingEntryESaIS6_EEERKNS5_INS_9AppMemoryESaISB_EEEbmb
            f340 _ZN15google_breakpad13WriteMinidumpEiiPKvmRKNSt7__cxx114listINS_12MappingEntryESaIS4_EEERKNS3_INS_9AppMemoryESaIS9_EEEbmb
            f380 _ZN15google_breakpad13WriteMinidumpEPKcliPKvmRKNSt7__cxx114listINS_12MappingEntryESaIS6_EEERKNS5_INS_9AppMemoryESaISB_EEEbmb
            f3c0 _ZN15google_breakpad13WriteMinidumpEiliPKvmRKNSt7__cxx114listINS_12MappingEntryESaIS4_EEERKNS3_INS_9AppMemoryESaIS9_EEEbmb
            f400 _ZN15google_breakpad17ProcCpuInfoReader12GetNextFieldEPPKc
            f660 _ZNSt6vectorIcN15google_breakpad16PageStdAllocatorIcEEE6resizeEm
            f8c0 _ZNSt6vectorI18MDMemoryDescriptorN15google_breakpad16PageStdAllocatorIS0_EEE19_M_emplace_back_auxIIRKS0_EEEvDpOT_
            fb10 _ZN15google_breakpad10TypedMDRVAI8MDStringE20CopyIndexAfterObjectEjPKvm.isra.7.part.8
            fb40 _ZN15google_breakpad18MinidumpFileWriterC1Ev
            fb60 _ZN15google_breakpad18MinidumpFileWriter4OpenEPKc
            fbe0 _ZN15google_breakpad18MinidumpFileWriter7SetFileEi
            fc10 _ZN15google_breakpad18MinidumpFileWriter5CloseEv
            fc70 _ZN15google_breakpad18MinidumpFileWriterD2Ev
            fc90 _ZN15google_breakpad18MinidumpFileWriter8AllocateEm
            fd50 _ZN15google_breakpad18MinidumpFileWriter4CopyEjPKvl
            fe50 _ZN15google_breakpad18MinidumpFileWriter20CopyStringToMDStringEPKwjPNS_10TypedMDRVAI8MDStringEE
            ff50 _ZN15google_breakpad18MinidumpFileWriter20CopyStringToMDStringEPKcjPNS_10TypedMDRVAI8MDStringEE
           10050 _ZN15google_breakpad12UntypedMDRVA8AllocateEm
           100a0 _ZN15google_breakpad18MinidumpFileWriter11WriteStringEPKwjP20MDLocationDescriptor
           102e0 _ZN15google_breakpad18MinidumpFileWriter11WriteStringEPKcjP20MDLocationDescriptor
           10520 _ZN15google_breakpad12UntypedMDRVA4CopyEjPKvm
           105b0 _ZN15google_breakpad18MinidumpFileWriter11WriteMemoryEPKvmP18MDMemoryDescriptor
           10690 _ZN15google_breakpad15UTF8ToUTF16CharEPKciPt
           10750 _ZN15google_breakpad16UTF32ToUTF16CharEwPt
           107d0 _ZN15google_breakpad11UTF16ToUTF8B5cxx11ERKSt6vectorItSaItEEb
           10ce0 _ZN15google_breakpad12UTF32ToUTF16EPKwPSt6vectorItSaItEE
           10ec0 _ZN15google_breakpad11UTF8ToUTF16EPKcPSt6vectorItSaItEE
           110a0 _ZNSt6vectorItSaItEE17_M_default_appendEm
           11200 _ZN15google_breakpad10IsValidElfEPKv
           11220 _ZN15google_breakpad8ElfClassEPKv
           11230 _ZN15google_breakpad14FindElfSectionEPKvPKcjPS1_Pm
           116a0 _ZN15google_breakpad15FindElfSegmentsEPKvjPNS_15wasteful_vectorINS_10ElfSegmentEEE
           11990 _ZNSt6vectorIN15google_breakpad10ElfSegmentENS0_16PageStdAllocatorIS1_EEE19_M_emplace_back_auxIIRKS1_EEEvDpOT_
           11be0 _ZN15google_breakpad6FileIDC2EPKc
           11cc0 _ZN15google_breakpad6FileID29ConvertIdentifierToUUIDStringB5cxx11ERKNS_15wasteful_vectorIhEE
           11e70 _ZN15google_breakpad6FileID25ConvertIdentifierToStringB5cxx11ERKNS_15wasteful_vectorIhEE
           11fb0 _ZN15google_breakpad6FileID31ElfFileIdentifierFromMappedFileEPKvRNS_15wasteful_vectorIhEE
           12bc0 _ZN15google_breakpad6FileID17ElfFileIdentifierERNS_15wasteful_vectorIhEE
           12c40 _Z10CreateGUIDP6MDGUID
           12d00 _Z12GUIDToStringPK6MDGUIDPci
           12d80 _ZN13GUIDGenerator12InitOnceImplEv
           12db0 my_strlen
           12dd0 my_strcmp
           12e20 my_strncmp
           12e80 my_strtoui
           12ed0 my_uint_len
           12f10 my_uitos
           12f50 my_strchr
           12f90 my_strrchr
           12fc0 my_memchr
           13000 my_read_hex_ptr
           13080 my_read_decimal_ptr
           130c0 my_memset
           130e0 my_strlcpy
           13120 my_strlcat
           13160 my_isspace
           13210 _ZN15google_breakpad16MemoryMappedFileC2Ev
           13220 _ZN15google_breakpad16MemoryMappedFile5UnmapEv
           13270 _ZN15google_breakpad16MemoryMappedFileD2Ev
           13280 _ZN15google_breakpad16MemoryMappedFile3MapEPKcm
           13430 _ZN15google_breakpad16MemoryMappedFileC1EPKcm
           13450 _ZN15google_breakpad12SafeReadLinkEPKcPcm
           13490 _ZN15google_breakpad12_GLOBAL__N_125CrashGenerationClientImplD1Ev
           134a0 _ZN15google_breakpad12_GLOBAL__N_125CrashGenerationClientImpl11RequestDumpEPKvm
           136c0 _ZN15google_breakpad12_GLOBAL__N_125CrashGenerationClientImplD0Ev
           136d0 _ZN15google_breakpad21CrashGenerationClient9TryCreateEi
           13700 _ZNK15google_breakpad10ThreadInfo21GetInstructionPointerEv
           13710 _ZNK15google_breakpad10ThreadInfo14FillCPUContextEP17MDRawContextAMD64
           13a10 _ZN15google_breakpad10ThreadInfo26GetGeneralPurposeRegistersEPPvPm
           13a60 _ZN15google_breakpad10ThreadInfo25GetFloatingPointRegistersEPPvPm
           13ab0 _ZN15google_breakpad14UContextReader15GetStackPointerEPK8ucontext
           13ac0 _ZN15google_breakpad14UContextReader21GetInstructionPointerEPK8ucontext
           13ad0 _ZN15google_breakpad14UContextReader14FillCPUContextEP17MDRawContextAMD64PK8ucontextPK13_libc_fpstate
           13d30 ConvertUTF32toUTF16
           13eb0 ConvertUTF16toUTF32
           14050 ConvertUTF16toUTF8
           14250 isLegalUTF8Sequence
           14330 ConvertUTF8toUTF16
           14660 ConvertUTF32toUTF8
           14920 ConvertUTF8toUTF32
           14c30 __libc_csu_init
           14ca0 __libc_csu_fini
           14ca4 _fini

